### DISCUSSION TITLE
Как обновить установленный Zapret?

### DISCUSSION THREAD

**Discussion by XRunner555** [bol-van/zapret#401] [General]
Думаю, что этот вопрос возникал у многих пользователей Zapret`a
Подскажите, пожалуйста, как обновить установленный пакет Zapret`a?
Остановка службы с последующей записью с заменой в /opt/zapret/ содержимого архива zapret-master.zip, сформированного на главной странице проекта, вызывает ошибку:  Permission denied и через стандартный способ:  /opt/etc/init.d/S90-zapret start служба после такого обновления не стартует.
Что делать? Как правильно обновлять установленный Запрет?

**Comment by bol-van** [comment]
Скачать архив в /tmp, распаковать , запустить /tmp/zapret-master/install_easy.sh, удалить /tmp/zapret-master

**Reply by bol-van** [reply]
конфиг сохраняется и важные файлы. перепись всего нужна, чтобы если в новой версии нет каких-то файлов, они удалились из /opt/zapret

**Reply by epetrovrbc** [reply]
> Скачать архив в /tmp, распаковать , запустить /tmp/zapret-master/install_easy.sh, удалить /tmp/zapret-master

А а если установлен через `git clone`, обновлять через `git pull` - это неправильно? Что-то сломается?

**Comment by XRunner555** [comment]
Офигеть, я конечно, очень тупой, но пользуюсь Линуксом только на роутере, пожалуйста, дайте перечень команд, которые нужно ввести после помещения содержимого архива  zapret-master.zip в директорию \opt\tmp\
И второй вопрос, службу запрета перед этим нужно останавливать?

**Reply by bol-van** [reply]
установить unzip 1 раз
opkg update
opkg install unzip

cd /tmp
unzip zapret-master.zip
zapret-master/install_easy.sh
.............. установка
rm -r /tmp/zapret-master

**Reply by XRunner555** [reply]
Спасибо, от души!

**Reply by XRunner555** [reply]
При запуске install_easy.sh весь процесс установки повторяется, все настройки заменяются на дефолные, а нет ли способа обновить только сервис, не проходя через процесс установки, как впервые?

**Reply by bol-van** [reply]
Конфиг и некоторые другие важные файлы должны сохраняться, если явно от этого не отказались

**Reply by XRunner555** [reply]
ОК, понял, спасибо за ответ.

**Comment by artemklevtsov** [comment]
Предлагаю свой вариант обновления на OpenWRT. Делал для себя, на универсальность не претендую.

Было бы здорово, если бы @bol-van включил нечто подобное в репозиторий.

Предполагается, что Zapret уже установлен.

```bash
mkdir -p /usr/local/bin/
cat <<EOF > /usr/local/bin/zapret-update.sh
#!/bin/sh

set -e

cd /tmp

ZAPRET_DIR=/opt/zapret
ZAPRET_BACKUP_DIR=/tmp/zapret-backup/
ZAPRET_TEMP_DIR=/tmp/zapret-master/
ZAPRET_URL=https://github.com/bol-van/zapret/archive/refs/heads/master.tar.gz
ZAPRET_TARBALL=/tmp/zapret.tar.gz

create_backup() {
  mkdir -p ${ZAPRET_BACKUP_DIR}
  mkdir -p ${ZAPRET_BACKUP_DIR}/ipset
  mkdir -p ${ZAPRET_BACKUP_DIR}/init.d/openwrt/custom.d

  cp ${ZAPRET_DIR}/config ${ZAPRET_BACKUP_DIR}/ 2> /dev/null
  cp ${ZAPRET_DIR}/update.sh ${ZAPRET_BACKUP_DIR}/ 2> /dev/null
  cp ${ZAPRET_DIR}/ipset/*.txt ${ZAPRET_BACKUP_DIR}/ipset/ 2> /dev/null
  cp ${ZAPRET_DIR}/init.d/openwrt/custom.d/* ${ZAPRET_BACKUP_DIR}/init.d/openwrt/custom.d/ 2> /dev/null
}

restore_backup() {
  mkdir -p ${ZAPRET_DIR}
  mkdir -p ${ZAPRET_DIR}/ipset

  cp -rv ${ZAPRET_BACKUP_DIR}/* ${ZAPRET_DIR}/
}

download_tarball() {
  wget -q -O ${ZAPRET_TARBALL} ${ZAPRET_URL}
}

extract_tarball() {
  tar -xzf ${ZAPRET_TARBALL}
  rm -f ${ZAPRET_TARBALL}
}

install_zapret() {
  rm -rf ${ZAPRET_DIR}
  mv ${ZAPRET_TEMP_DIR} ${ZAPRET_DIR}
  sh ${ZAPRET_DIR}/install_bin.sh
}

post_install_zapret() {
  sh ${ZAPRET_DIR}/ipset/clear_lists.sh
  sh ${ZAPRET_DIR}/ipset/get_config.sh
}

[[ ! -d ${ZAPRET_DIR} ]] && exit 0

echo "Stop service"
service zapret stop

echo "Download zapret tarball from GitHub"
download_tarball

echo "Extract zapret tarball"
extract_tarball

echo "Update zapret files"
create_backup
install_zapret
restore_backup
post_install_zapret

echo "Start service"
service zapret start
EOF

chmod +x /usr/local/bin/zapret-update.sh

cat <<EOF >> /etc/sysupgrade.conf
/opt/zapret/config
/opt/zapret/ipset/zapret-hosts-user.txt
/opt/zapret/ipset/zapret-hosts-user-exclude.txt
/opt/zapret/init.d/openwrt/custom.d/
/usr/local/bin/
EOF

cat <<EOF >> /root/.profile
export PATH=$PATH:/usr/local/bin
EOF
```

Теперь достаточно вызвать:

```bash
zapret-update.sh
```

**Reply by remittor** [reply]
> Предлагаю свой вариант обновления на OpenWRT.

Вот мой вариант: https://github.com/remittor/zapret-openwrt

При обновлении текущие конфиги не заменяются.
Но в основной uci конфиг добавляются новые параметры (если они есть).

При удалении конфиги остаются.
При бэкапе всех настроек openwrt в tgz файл складируются все конфиги zapret.

**Reply by disinformer** [reply]
Но надо же предварительно установите opkg-репозиторий в систему, верно?

**Reply by ShNURoK42** [reply]
> > Предлагаю свой вариант обновления на OpenWRT.
> 
> Вот мой вариант: https://github.com/remittor/zapret-openwrt
> 
> При обновлении текущие конфиги не заменяются. Но в основной uci конфиг добавляются новые параметры (если они есть).
> 
> При удалении конфиги остаются. При бэкапе всех настроек openwrt в tgz файл складируются все конфиги zapret.

Напишите пожалуйста как установить, кто не работал с opkg. И надо ли удалять установленную руками версию?

**Reply by remittor** [reply]
> И надо ли удалять установленную руками версию?

Надо. Если не удалите, то мой модуль при установке отругнётся и установка прервётся.

> Напишите пожалуйста как установить, кто не работал с opkg.

На openwrt работа с opkg может производится через WEB. Сами попробуйте найти, где это расположено.

**Comment by chipodrillo** [comment]
Приветствую. А вообще можно ли как то посмотреть какая сейчас установленна версия "Zapret" ? (Роутер с OpenWRT)

**Reply by numbereleven-a** [reply]
> Приветствую. А вообще можно ли как то посмотреть какая сейчас установленна версия "Zapret" ? (Роутер с OpenWRT)

Косвенно только если, можно узнать меньше 66ой или больше.
Начиная с 66ой в конфиге настройка, например NFQWS_OPT_DESYNC_HTTP заменена на NFQWS_OPT="
--filter-tcp=80

**Comment by chipodrillo** [comment]
Здравствуйте. Подскажите пожалуйста, можноли скачать архив с релиза прям на роутер ? Достали эти антивирусы...

**Reply by chipodrillo** [reply]
Пример:
wget https://github.com/bol-van/zapret/releases/download/v70.6/zapret-v70.6.zip
качает с неправельным именем файла... Руками попровляю на zip ... Подскажите как правельно прописывать...?, чтобы сразу прапвельное имя файла.. "zapret-v70.6.zip"

**Reply by chipodrillo** [reply]
> > Подскажите пожалуйста, можноли скачать архив с релиза прям на роутер ? Достали эти антивирусы...
> 
> Вот отсюда можете скачать: https://github.com/remittor/zapret-openwrt/releases На эти модули антивири не орут.

Да хотелось бы с офф источника... Всё таки так надёжней и правельней.

Разобрался:
wget https://github.com/bol-van/zapret/releases/download/v70.6/zapret-v70.6.zip -O /tmp/zapret-v70.6.zip

**Comment by Dmitry-77** [comment]
Вопрос от полного ламера.
Asus rt-n56u, прошивка padavan 3.4.3.9L-101_ed35e25 с запретом github version v71.1.1.
При попытке установить zapret-v71.4.zip или zapret-v71.4-openwrt-embedded.tar.gz
получаю:
```
* checking system
zap/install_easy.sh: line 184: readlink: not found
zap/install_easy.sh: line 201: readlink: not found
system is not either systemd, openrc or openwrt based
easy installer can set up config settings but can't configure auto start
you have to do it manually. check readme.md for manual setup info.
```
Что я не так делаю, и где взять zapret-master.zip?

**Reply by chipodrillo** [reply]
> Вопрос от полного ламера. Asus rt-n56u, прошивка padavan 3.4.3.9L-101_ed35e25 с запретом github version v71.1.1. При попытке установить zapret-v71.4.zip или zapret-v71.4-openwrt-embedded.tar.gz получаю:
> 
> ```
> * checking system
> zap/install_easy.sh: line 184: readlink: not found
> zap/install_easy.sh: line 201: readlink: not found
> system is not either systemd, openrc or openwrt based
> easy installer can set up config settings but can't configure auto start
> you have to do it manually. check readme.md for manual setup info.
> ```
> 
> Что я не так делаю, и где взять zapret-master.zip?

Скачай релиз *zip и вытащи из архива

**Reply by Dmitry-77** [reply]
Так вытащил же. Об line 184 и line 201 споткнулся инсталлятор. "system is not either systemd, openrc or openwrt based"

**Reply by kost2000** [reply]
Насколько я понимаю, так не обновить Zapret. Надо заново пересобрать прошивку. 

Если есть USB, то можно поставить  Entware. А там уже установить Zapret. Я пользовался этим пакетом - https://github.com/Anonym-tsk/nfqws-keenetic

**Comment by Dmitry-77** [comment]
На арчере с5 обновил прошивку, а с ней и запрет обновился до 71.4.
[Проблему с отвалившимся ютубом](https://github.com/bol-van/zapret/discussions/200#discussioncomment-14372053) и пр. это не решило.

**Comment by Dmitry-77** [comment]
Вот так обновляется zapret на роутерах с прошивкой padavan
`zapret.sh download && zapret.sh restart`
Обновилось на актуальную 71.4. [Взято здесь.](https://4pda.to/forum/index.php?showtopic=428005&view=findpost&p=136770153)

**Reply by andmailg** [reply]
Благодарю!
Конфиги сохранились после обновления или заменились на дефолтные?

**Reply by Dmitry-77** [reply]
Конфиги сохранились.

**Comment by standhaftsohnsergius** [comment]
cd /opt/zapret && \
 cp config /root/zapret-config-backup-$(date +%Y%m%d).txt && \
 git pull origin master && \
 ./install_bin.sh && \
 systemctl restart zapret && \
 journalctl -u zapret -n 20
